### PR TITLE
BUG: Fix RGB FlateEncode Images(PNG) and transparency

### LINKED ISSUE #1368: Images not extracted
When trying to extract the images, PyPDF2 didn't capture them.

## Environment

```bash
$ python -c "import PyPDF2;print(PyPDF2.__version__)"
2.11.0
```

## Code + PDF

Use this PDF: https://arxiv.org/pdf/2201.00151.pdf

```python
>>> from PyPDF2 import PdfReader
>>> reader = PdfReader("2201.00151.pdf")
>>> reader.pages[1].images
[]
```

Expected were two images.

### LINKED ISSUE #1599: Images with transparency mask are not correctly extracted
When an image is extracted

```python
from pypdf import PdfReader, PdfWriter
from pypdf.generic import NameObject, NullObject
from PIL import Image
from io import BytesIO

w = PdfWriter()
w.append("resources/labeled-edges-center-image.pdf")

for p in w.pages:
    for image_file_object in p.images:
        print(image_file_object.name)
        ii = Image.open(BytesIO(image_file_object.data))
        b = BytesIO()
        ii.save(b, "pdf", quality=60, resolution=19.0, optimize=True)
        rrr = PdfReader(b)
        n = NameObject("/" + "".join(image_file_object.name.split(".")[:-1]))
        ind = p["/Resources"]["/XObject"].raw_get(n)
        w._objects[ind.idnum] = NullObject()  # to cleanup file
        p["/Resources"]["/XObject"][n] = (
            rrr.pages[0]["/Resources"]["/XObject"]["/image"].clone(w).indirect_reference
        )
w.write("tt.pdf")
```

edit : code updated

_Originally posted by @pubpub-zz in https://github.com/py-pdf/pypdf/issues/1546#issuecomment-1411144670_

### LINKED ISSUE #1787: pypdf.errors.PdfReadError: Image data is not rectangular
Hi guys, how are you? First of all thank you so much for making and mantaining this amazing library!
I want to extract all images in the pdf, but failed.
I cannot confirm whether it is inline image or not, I tried to compress the pdf ,then read the orginal instruction sequences, but I couldn't find the 'BI' tag

## Environment

Which environment were you using when you encountered the problem?

```bash
$ python -m platform
Windows-10-10.0.19041-SP0


$ python -c "import pypdf;print(pypdf.__version__)"
3.6.0
```

## Code + PDF

This is a minimal, complete example that shows the issue:

```python
from pypdf import PdfReader

reader = PdfReader("pdf_font_garbled.pdf")
a=reader.pages[1].images
```

Share here the PDF file(s) that cause the issue. The smaller they are, the
better. Let us know if we may add them to our tests!
[pdf_font_garbled.pdf](https://github.com/py-pdf/pypdf/files/11219022/pdf_font_garbled.pdf)

## Traceback

This is the complete Traceback I see:

```
Traceback (most recent call last):
  File "D:\code_project\try_pdf\experi\pypdf_image.py", line 15, in <module>
    a=reader.pages[1].images
  File "D:\code_project\venv_all\try_pdf_latest\lib\site-packages\pypdf\_page.py", line 444, in images
    extension, byte_stream = _xobj_to_image(x_object[obj])
  File "D:\code_project\venv_all\try_pdf_latest\lib\site-packages\pypdf\filters.py", line 600, in _xobj_to_image
    data = x_object_obj.get_data()  # type: ignore
  File "D:\code_project\venv_all\try_pdf_latest\lib\site-packages\pypdf\generic\_data_structures.py", line 882, in get_data
    decoded._data = decode_stream_data(self)
  File "D:\code_project\venv_all\try_pdf_latest\lib\site-packages\pypdf\filters.py", line 545, in decode_stream_data
    data = FlateDecode.decode(data, stream.get(SA.DECODE_PARMS))
  File "D:\code_project\venv_all\try_pdf_latest\lib\site-packages\pypdf\filters.py", line 145, in decode
    str_data = FlateDecode._decode_png_prediction(str_data, columns, rowlength)  # type: ignore
  File "D:\code_project\venv_all\try_pdf_latest\lib\site-packages\pypdf\filters.py", line 156, in _decode_png_prediction
    raise PdfReadError("Image data is not rectangular")
pypdf.errors.PdfReadError: Image data is not rectangular
```

### LINKED ISSUE #1863: bug : Issues during image extraction
from @ericgonzadev

### Environment
Mac OS 12.6.6

```
$ python3 -c "import pypdf;print(pypdf.__version__)"
3.9.0
```

### Code + PDF
```
from pypdf import PdfReader
reader = PdfReader(pdf_file)
for index, page in enumerate(reader.pages):
     for img in page.images:
        print(f"{index}-{img.name}")
```

Tested with the following pdf: https://ufile.io/o1whh9b3
The error seems to be happening on page # 34 

### Traceback
```
Traceback (most recent call last):
  File "/Users/eric/Downloads/pdf_extract_images.py", line 60, in scan_directory
    for img in page.images:
                      ^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pypdf/_page.py", line 463, in images
    extension, byte_stream = _xobj_to_image(x_object[obj])
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pypdf/filters.py", line 707, in _xobj_to_image
    alpha = Image.frombytes("L", size, x_object_obj[G.S_MASK].get_data())
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/PIL/Image.py", line 2970, in frombytes
    im.frombytes(data, decoder_name, args)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/PIL/Image.py", line 826, in frombytes
    raise ValueError(msg)
ValueError: not enough image data
```

_Originally posted by @ericgonzadev in https://github.com/py-pdf/pypdf/issues/1814#issuecomment-1562053211_

### PR DESCRIPTION
**Pull request by pubpub-zz**
Take the number of colors into account for PNG images

Properly process the mask for transparency:

Fixes #1787
Fixes #1599

Adds support for inline images extraction:

Fixes #1368
Fixes #1863

Additional changes:

* Process TIFF predictor 2
* Upgrades Pillow requirement to version 9.5 for Python 9.11

### CODE REVIEW

**Comment by pubpub-zz** [comment]
test images ;
from resources/labeled-edges-center-image.pdf  (PNG + transparency wirht reversed scale)
labeled

from [pdf_font_garbled.pdf](https://github.com/py-pdf/pypdf/files/11219022/pdf_font_garbled.pdf):
2nd page watermark (RGB+mask PNG):
i0

Image45 part of resources in TPL2 in 2nd page (jp2 file => zipped to stored)  (JPG + transparency => JP2)
[p65.zip](https://github.com/py-pdf/pypdf/files/11412697/p65.zip)

**Comment by pubpub-zz** [comment]
tiff predictor 2:
from https://corpora.tika.apache.org/base/docs/govdocs1/977/977609.pdf
image :
pdftiffimage

File: pypdf/_page.py [hunk 1]
```diff
@@ -460,12 +460,94 @@ def images(self) -> List[File]:
...
+            Read-only property that emulates a list of files
+            Get a list of all images of the page.
+
+            the key can be:
+              µan str (for top object) or a tuple for image within XObject forms
+              or an int
+        ex:
+        ```
+        reader.pages[0].images[0]        # return fist image
+        reader.pages[0].images['/I0']    # return image '/I0'
```
Comment (MartinThoma) [inline pypdf/_page.py:532]: Can it happen that there are name clashes, e.g. that there are two different images with the same key?

File: pypdf/filters.py [hunk 2]
```diff
@@ -695,19 +718,36 @@ def _xobj_to_image(x_object_obj: Dict[str, Any]) -> Tuple[Optional[str], bytes]:
                 mode_map = {
                     "/DeviceGray": "L",
                     "/DeviceRGB": "RGB",
-                    "/DeviceCMYK": "RGBA",
+                    "/DeviceCMYK": "CMYK",  # used to be "RGBA" but this is seems not in accordance withFlateEncode Spec
```
Comment (MartinThoma) [inline pypdf/filters.py:721]: I agree with this change - I don't think we need the comment here

```suggestion
                    "/DeviceCMYK": "CMYK",
```

**Comment by pubpub-zz** [comment]
images possible updates
i0
Im0

File: pypdf/constants.py [hunk 3]
```diff
@@ -451,6 +451,7 @@ class GraphicsStateParameters:
     SM = "/SM"
     SA = "/SA"
     BM = "/BM"
+    MASK = "/Mask"  # 1-bit image mask stream
```
Comment (MartinThoma) [inline pypdf/constants.py:454]: That does not belong here, but into the `ImageAttributes` class (PDF 1.7: "Table 89 – Additional Entries Specific to an Image Dictionary" or PDF 2.0: "Table 87 — Additional entries specific to an image dictionary")

**Comment by pubpub-zz** [comment]
new test file (CMYK) with transparency
eole

File: pypdf/constants.py (hunk 3 above)
Comment (pubpub-zz) [reply pypdf/constants.py]: thanks : fixed

File: pypdf/_page.py (hunk 1 above)
Comment (pubpub-zz) [reply pypdf/_page.py:546]: no : it is a dictionary actually.
The only case you can get is that the name is identical within two XObject/Forms with different names (the tuple / sequence will be different)

File: pypdf/filters.py (hunk 2 above)
Comment (pubpub-zz) [reply pypdf/filters.py]: applied

**Comment by MartinThoma** [comment]
@pubpub-zz This PR introduces two changes: (1) The bugfix and (2) the new interface.

Could you please make a second PR just with the bugfix? I feel like that one can be merged today/tomorrow, but I want to have another look at the new interface.

**Comment by pubpub-zz** [comment]
> @pubpub-zz This PR introduces two changes: (1) The bugfix and (2) the new interface.
> 
> Could you please make a second PR just with the bugfix? I feel like that one can be merged today/tomorrow, but I want to have another look at the new interface.

For tracking I've isolated the new images and 2 other PR following : inline_images and new function to replace images. However I'm not inclined to process only the bugs ones appart : I've worked all changes at the same time and I would not like to introduce regressions with a mod not being correctly reported.

**Comment by MartinThoma** [comment]
@pubpub-zz I messed something up when I fixed the merge conflicts :see_no_evil: 

```
__________________________________ test_cmyk ___________________________________

    @pytest.mark.enable_socket()
    def test_cmyk():
        """Decode cmyk with transparency"""
        url = "https://corpora.tika.apache.org/base/docs/govdocs1/972/972174.pdf"
        name = "tika-972174.pdf"
        reader = PdfReader(BytesIO(get_pdf_from_url(url, name=name)))
        url_png = "https://user-images.githubusercontent.com/4083478/238288207-b77dd38c-34b4-4f4f-810a-bf9db7ca0414.png"
        name_png = "tika-972174_p0-im0.png"
        refimg = Image.open(
            BytesIO(get_pdf_from_url(url_png, name=name_png))
        )  # not a pdf but it works
        data = reader.pages[0].images[0]
        assert ".jp2" in data.name
>       assert list(data.image.getdata()) == list(refimg.getdata())
E       assert [(255, 255, 255, 0),\n (255, 255, 255, 0),\n (255, 255, 255,
```

**Comment by MartinThoma** [comment]
@pubpub-zz Can you please help me? I have no idea where to look :see_no_evil:

**Comment by pubpub-zz** [comment]
under analysis.

**Comment by pubpub-zz** [comment]
shoud  be fixed. I had to upgrade pillow to 9.5 for python 3.11

**Comment by MartinThoma** [comment]
Thank you :pray: 

I'm quickly checking if any of the other PRs is ready to merge and will make a release in less than an hour. This time we have awesome improvements for image handling thanks to your efforts. That was really good work :+1: :hugs:

**Pull request merged**